styx-scheduler-service/src/main/java/com/spotify/styx/state/StateUtil.java

```diff
@@ -0,0 +1,166 @@
+                                                                 List<InstanceState> activeStates,
+                                                                 Set<WorkflowInstance> timedOutInstances,
+                                                                 WorkflowResourceDecorator resourceDecorator,
+                                                                 Set<Workflow> workflows) {
+    return activeStates.parallelStream()
+        .filter(entry -> !timedOutInstances.contains(entry.workflowInstance()))
+        .filter(entry -> isConsumingResources(entry.runState().state()))
+        .flatMap(instanceState -> pairWithResources(globalConcurrencyEnabled, instanceState,
+            workflows, resourceDecorator))
+        .collect(groupingByConcurrent(
+            ResourceWithInstance::resource,
+            ConcurrentHashMap::new,
+            counting()));
+  }
+
+  public static Map<String, Long> getResourcesUsageMap(Storage storage, TimeoutConfig timeoutConfig,
+                                                       WorkflowCache workflowCache, Instant instant,
+                                                       WorkflowResourceDecorator resourceDecorator)
+      throws IOException {
+    // The only inconsistency left is to miss active workflow instances. Outdated instances or
+    // inactive instances will be correctly updated or removed via the strongly consistent lookups
+    final Map<WorkflowInstance, RunState> strictActiveStates = storage.readActiveStates().entrySet()
+        .parallelStream().filter(entry -> {
+          try {
+            return storage.readActiveState(entry.getKey()).isPresent();
+          } catch (IOException e) {
+            throw new RuntimeException("Error while fetching active states", e);
+          }
+        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    final List<InstanceState> activeInstanceStates = getActiveInstanceStates(strictActiveStates);
+    boolean globalConcurrencyEnabled = storage.config().globalConcurrency().isPresent();
+    final Set<WorkflowInstance> timedOutInstances =
+        getTimedOutInstances(activeInstanceStates, instant, timeoutConfig);
+    return getResourceUsage(globalConcurrencyEnabled,
+        activeInstanceStates, timedOutInstances, resourceDecorator, workflowCache.all());
+  }
+
+  private static Stream<ResourceWithInstance> pairWithResources(boolean globalConcurrencyEnabled,
+                                                                          InstanceState instanceState,
+                                                                          Set<Workflow> workflows,
+                                                                          WorkflowResourceDecorator resourceDecorator) {
+    final Optional<Workflow> workflowOpt = workflows.stream().filter(
+        wf -> wf.id().equals(instanceState.workflowInstance().workflowId())).findFirst();
```

Comment:
Just pass in a map instead and do `O(1)` lookups instead of linear search.

Reply:
Exactly. Will fix it in https://github.com/spotify/styx/pull/432